L67-cpp-callback-api.md

```diff
@@ -0,0 +1,126 @@
+   - `Finish`: for server RPCs to provide completion status. This initiates the asynchronous transmission of the status to the client. Must be used for all server RPCs, including those that have already been cancelled. *Server only*.
+   - _Constraints_: There may not be more than one concurrent `StartRead` and more than one concurrent `StartWrite` in-progress for a given RPC. However, `StartRead` and `StartWrite` may be concurrent. These operations are considered done when the `OnReadDone` or `OnWriteDone` reactions is invoked. There must be exactly one call of `Finish` at the server and `StartCall` at the client. There may be no more than one use of `StartWritesDone` at the client or `StartSendInitialMetadata` at the server. Any invocation of `StartSendInitialMetadata` at the server must happen before an invocation of `StartWrite` or `Finish`.
+1. Operation-completion reaction methods: notification of completion of asynchronous RPC activity. These are all virtual methods that default to an empty function (i.e., `{}`) but may be overridden by the application's reactor definition.
+   - `OnReadDone`, `OnWriteDone`, `OnWritesDoneDone`: methods invoked by the library at the completion of an explicitly-specified streaming RPC operation. Provides a `bool ok` argument to indicate whether the completion was successful. *Client and server*. 
+   - `OnReadInitialMetadataDone`: notifying client RPCs that the server has sent initial metadata. It has a bool `ok` argument to indicate whether the completion was successful. This is automatically invoked by the library for all RPCs; it is uncommonly used however. *Client only*
+   - `OnCancel`: for server RPCs only, called if an RPC is canceled before it has a chance to successfully send status to the client side. Should be used for any cleanup associated with cancellation. Note that even cancelled RPCs _must_ get a `Finish` call to cleanup all their library state. *Server only*
+   - `OnDone`: called when all outstanding RPC operations are completed for both client and server RPCs. For the client-side, it additionally provides the status that the server sent with its `Finish` call. Should be used for any application-level RPC-specific cleanup. *Client and server.*
+   - _Thread safety_: the above calls may take place concurrently, except that `OnDone` will always take place after all other reactions. No further RPC operations should be issued after `OnDone` is invoked.
+   - **IMPORTANT USAGE NOTE** : code in any reaction must not block for an arbitrary amount of time since reactions are executed on a finite-sized library-controlled threadpool. If any long-term blocking operations (like sleeps, file I/O, synchronous RPCs, or waiting on a condition variable) must be invoked as part of the application logic, then it is important to push that outside the reaction so that the reaction can complete in a timely fashion. One way of doing this is to push that code to a separate application-controlled thread.
```

Comment:
When you were prototyping this inside of Google, did you ever emit a log message when a reaction blocked for a long time? What that useful? Did it end up in the gRPC version of the implementation?

I'm unsure if such a log message would be useful or if it would be noise. Is there a "good enough" static threshold?

Reply:
I should clarify that the Google version wasn't a separate prototype; the code that we've been running in Google is literally the callback API that has been developed out here at Github in bits and pieces since 2019. It was just usable internally because we had an additional  pre-existing piece that is hinted at in the implementation: an iomgr variant that satisifes `grpc_iomgr_run_in_background`. We had hoped to make an OSS version of that in #182 but had to abandon that project because of dependences, though it might be revived in yet another form down the road. That was why this ended up taking longer than planned and how we worked around that with grpc/grpc#25169 .

But getting back to your question: we never tried such a log but it seems like it would be a good idea. We could, for example, start a timer when a callback is initiated for say 100 ms in the future and then either cancel it at callback completion or let it fire and log. I suppose that this could be a useful debugging tool.

Reply:
> We could, for example, start a timer when a callback is initiated for say 100 ms in the future and then either cancel it at callback completion or let it fire and log. I suppose that this could be a useful debugging tool.

That would work. It may be even cheaper to just time how long the callback takes and log after it returns, even if it takes 12 minutes. That would reduce the overhead of having a timer and another callback.

```scheme
(let [(timer (start-timer))
      (callback-result (callback ...))]
  (begin
    (if (> (timer-duration timer) (ms 100)) (log-too-long ...))
    callback-result))
```

> We had hoped to make an OSS version of that in #182

I remember that proposal. The "handle" type in there really needs to be at least pointer sized to work on non-Linux OSes. The native implementations that one would likely use to implement an I/O manager on both Windows (I/O completion queues) and macOS (libdispatch) use pointer-sized handles, not `int`. In general, wherever Linux has `int fd`, it seems like gRPC will need need `void* handle` to work cross-platform.

Reply:
Are you taunting me with Lisp as a fellow Rice alum? Seems like a reasonable thing to do. We should add some plumbing to pass the callback's `__FILE__` and `__LINE__` into the library so that the log would be human-readable and useful.

#182 failed for reasons beside handlesize iirc

Reply:
> Are you taunting me with Lisp as a fellow Rice alum?

Yup! And, come now, it was Scheme. It had square brackets. 😜

Reply:
It's been 30 years since I took COMP 210 so you'll have to forgive my lapse. The book was called The Little LISPer back then even though it was Scheme.